src/wifi.rs

```diff
@@ -1442,3 +1445,144 @@ impl WifiWait {
+where
+    T: BorrowMut<WifiDriver<'d>>,
+{
+    base_driver: T,
+    _p: PhantomData<WifiDriver<'d>>,
+}
+
+#[cfg(feature = "nightly")]
+impl<'d, T> AsyncWifiDriver<'d, T>
+where
+    T: BorrowMut<WifiDriver<'d>>,
+{
+    #[cfg(all(feature = "alloc", esp_idf_comp_nvs_flash_enabled))]
+    pub fn new(base_driver: T) -> Result<Self, EspError> {
+        Ok(AsyncWifiDriver {
+            base_driver,
+            _p: PhantomData,
+        })
+    }
+
+    pub fn get_capabilities(&self) -> Result<EnumSet<Capability>, EspError> {
+        self.base_driver.borrow().get_capabilities()
+    }
+
+    pub fn get_configuration(&self) -> Result<Configuration, EspError> {
+        self.base_driver.borrow().get_configuration()
+    }
+
+    pub fn set_configuration(&mut self, conf: &Configuration) -> Result<(), EspError> {
+        self.base_driver.borrow_mut().set_configuration(conf)
+    }
+
+    pub fn is_started(&self) -> Result<bool, EspError> {
+        self.base_driver.borrow().is_started()
+    }
+
+    pub fn is_connected(&self) -> Result<bool, EspError> {
+        self.base_driver.borrow().is_connected()
+    }
+
+    pub async fn start(&mut self) -> Result<(), EspError> {
+        self.base_driver.borrow_mut().start()
+        // TODO: wait until start is complete
+    }
+
+    pub async fn stop(&mut self) -> Result<(), EspError> {
+        self.base_driver.borrow_mut().stop()
+        // TODO: wait until stop is complete?
+    }
+
+    pub async fn connect(&mut self) -> Result<(), EspError> {
+        self.base_driver.borrow_mut().connect()
+        // TODO: wait until connect is complete
+    }
+
+    pub async fn disconnect(&mut self) -> Result<(), EspError> {
+        self.base_driver.borrow_mut().disconnect()
+        // TODO: wait until stop is complete
+    }
+
+    pub async fn scan_n<const N: usize>(
+        &mut self,
+    ) -> Result<(heapless::Vec<AccessPointInfo, N>, usize), EspError> {
+        // TODO: rewrite to do async waiting here while scanning
+        // Also, it would be nice to have an API that returned entries as they were discovered
+        self.base_driver.borrow_mut().scan_n()
+    }
+
+    pub async fn scan(&mut self) -> Result<Vec<AccessPointInfo>, EspError> {
+        let base: &mut WifiDriver<'d> = self.base_driver.borrow_mut();
+        base.scan()
+    }
+}
+
+#[cfg(all(feature = "nighty", feature = "experimental"))]
+impl<'d, T> asynch::Wifi for AsyncWifiDriver<'d, T>
+where
+    T: BorrowMut<WifiDriver<'d>>,
+{
+    type Error = EspError;
+
+    type GetCapabilitiesFuture<'a> = futures::future::Ready<Result<EnumSet<Capability>, EspError>> where Self: 'a;
+    type GetConfigurationFuture<'a> = futures::future::Ready<Result<Configuration, EspError>> where Self: 'a;
+    type SetConfigurationFuture<'a> = futures::future::Ready<Result<(), EspError>> where Self: 'a;
+    type StartFuture<'a> = Pin<Box<dyn futures::future::Future<Output = Result<(), EspError>> + 'a>> where Self: 'a;
```

Comment:
Hint: here you need the TAIT future which is nightly only. No Pin, no Box. Basically, `= impl Future<Output = Result<(), Self::Error>>`.